### DISCUSSION TITLE
Input cleaned data to mtex for IPF mapping

### DISCUSSION THREAD

**Discussion by lebraunjames** [mtex-toolbox/mtex#1991] [Ask Anything] [answered]
Hello MTEX Community,

The IPF color mapping of mtex is so much more presentable than other ebsd processing software. However, my problem is I can't load a cleaned map with .h50ina file type, hence there are a lot of unindexed regions when i plot the IPF. I can't replicate the zero solution removal cleaning in aztec in mtex so i try to clean it before hand, but mtex disregards it and loads the unindexed points and not their cleaned counterparts. Is there a way to fix this?

Thank you so much for your guidance.

Regards,
LJ

**Comment by kilir** [comment, marked as answer]
Hi,

you can check out the lastest version  f993c6c009d0a20604c786eaecd2b1ea4f02e5ac 
With ebsd = loadEBSD_h5oina('filename.h5oina','useProcessedData'), it should use the processed data (with the exception of grainId at the moment).

Anyhow, which cleanup steps do you think you can only accomplish in aztec crystal but not in mtex?

Cheers,
Rüdiger

**Comment by lebraunjames** [comment]
Thankyou so much for the update! I'm excited about the IPFs that i can generate from this one :). As for the clean up, i'm not sure if iterating the nearest neighbors function can be replicated right now in mtex. I tried iterating these functions but it never seems to fully fill the unindexed regions:
```matlab
ebsd_filled = fill(ebsd,grains)
```
and
```matlab
F = halfQuadraticFilter;
F.alpha = 0.25;

ebsdSub_filled = smooth(ebsd,F,'fill',grains);
```
Again, i am so grateful for the update. Cheers!

**Reply by kilir** [reply]
Hi,
the reason is that there's a difference between points which are not in the list of points and explicitly notIndexed points.

Have a look at the `erode` function which does (iterative) erosion of points based on the number of neighbors. This, in combination with `EBSD/fill` allows for many possibilities.

Examples (white pixels are the ones that do not exist anymore and can potentially be filled):
```
mtexdata small
plot(ebsd); hold on
plot(ebsd('n'),'FaceColor','k'); hold off;
mtexTitle('original');  nextAxis

% erode isolated, nonIndexed points
ebsd = erode(ebsd,0);
plot(ebsd); hold on
plot(ebsd('n'),'FaceColor','k'); hold off;
mtexTitle('no single nonIndexed pixels');  nextAxis

% erode isolated, nonIndexed points
for i=1:10
ebsd = erode(ebsd,1,{'n' 'D' 'E'});
end

plot(ebsd); hold on
plot(ebsd('n'),'FaceColor','k'); hold off;
mtexTitle('no n,D,E $\le$ 1 neighbors, 10 iter'); nextAxis


% erode isolated, nonIndexed points

for i=1:10
ebsd = erode(ebsd,2,{'n' 'D' 'E','F'});
ebsd = ebsd.fill;
end

plot(ebsd); hold on
plot(ebsd('n'),'FaceColor','k'); hold off;
mtexTitle('no n,D,E $\le$ 2 neighbors, filled,10 iter'); 
```
Cheers,
Rüdiger

**Reply by lebraunjames** [reply]
Hello,

Thanks for letting me know. I tried your code and it works with the example, but for some reason it doesn't work in my raw dataset. Can you spot something wrong with this code? I can also share the raw data if needed.

```
% crystal symmetry
CS = {... 
  'notIndexed',...
  crystalSymmetry('m-3m', [3.7 3.7 3.7], 'mineral', 'Iron fcc', 'color', [0.53 0.81 0.98]),...
  crystalSymmetry('m-3m', [2.9 2.9 2.9], 'mineral', 'Iron bcc (old)', 'color', [0.56 0.74 0.56])};

% path to files
pname = 'C:\Users\...';

% which files to be imported
fname = [pname '\dataset.h5oina'];

% create an EBSD variable containing the data
ebsd = loadEBSD_h5oina(fname);

ebsd = erode(ebsd,0);
for i=1:10
ebsd = erode(ebsd,1,{'Iron bcc (old)'});
end
for i=1:50
ebsd = erode(ebsd,3,{'Iron bcc (old)'});
end

% identify grains (made super small to include more data)
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'angle',3*degree);
grains = smooth(grains,5);

ipfKey = ipfHSVKey(ebsd('Iron bcc (old)'));
ipfKey.inversePoleFigureDirection = vector3d.Z;
plotx2east
plotzIntoPlane
plot(ebsd('Iron bcc (old)'),ebsd('Iron bcc (old)').orientations,'figSize','large','micronbar','off')
hold on
plot(grains.boundary,'linewidth',1)
hold off
```


Somehow unindexed points remain unfilled after this. 
Cheers,
LJ
